runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.beam.sdk.util.state.StateInternalsFactory;
+import org.apache.beam.sdk.util.state.TimerInternalsFactory;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+class SplittableProcessElementsEvaluatorFactory<InputT, OutputT, RestrictionT>
+    implements TransformEvaluatorFactory {
+  private final ParDoEvaluatorFactory<
+          KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>, OutputT>
+      delegateFactory;
+  private final EvaluationContext evaluationContext;
+
+  SplittableProcessElementsEvaluatorFactory(EvaluationContext evaluationContext) {
+    this.evaluationContext = evaluationContext;
+    this.delegateFactory = new ParDoEvaluatorFactory<>(evaluationContext);
+  }
+
+  @Override
+  public <T> TransformEvaluator<T> forApplication(
+      AppliedPTransform<?, ?, ?> application, CommittedBundle<?> inputBundle) throws Exception {
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    TransformEvaluator<T> evaluator =
+        (TransformEvaluator<T>)
+            createEvaluator((AppliedPTransform) application, (CommittedBundle) inputBundle);
+    return evaluator;
+  }
+
+  @Override
+  public void cleanup() throws Exception {
+    delegateFactory.cleanup();
+  }
+
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private TransformEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>
+      createEvaluator(
+          AppliedPTransform<
+                  PCollection<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>,
+                  PCollectionTuple, SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT>>
+              application,
+          CommittedBundle<InputT> inputBundle)
+          throws Exception {
+    final SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT> transform =
+        application.getTransform();
+
+    DoFnLifecycleManager fnManager = delegateFactory.getFnClone(transform.getProcessFn());
+
+    SplittableParDo.ProcessFn<InputT, OutputT, RestrictionT, ?> processFn =
+        ((SplittableParDo.ProcessFn<InputT, OutputT, RestrictionT, ?>) fnManager.get());
+
+    String stepName = evaluationContext.getStepName(application);
+    final DirectExecutionContext.DirectStepContext stepContext =
+        evaluationContext
+            .getExecutionContext(application, inputBundle.getKey())
+            .getOrCreateStepContext(stepName, stepName);
+
+    ParDoEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>, OutputT>
+        parDoEvaluator =
+            delegateFactory.createParDoEvaluator(
+                application,
+                transform.getSideInputs(),
+                transform.getMainOutputTag(),
+                transform.getSideOutputTags().getAll(),
+                stepContext,
+                fnManager);
+
+    processFn.setStateInternalsFactory(
```

Comment:
OK, I changed the code to create a ProcessFn at evaluation time. However, I still can't pass StateInternals, TimerInternals and OutputWindowedValue in the constructor due to cyclic dependencies between these and the surrounding objects - e.g. DoFnTester (e.g. the ProcessFn depends on DoFnTester.getTimerInternals(), and DoFnTester obviously depends on the ProcessFn) and ParDoEvaluator (similar cyclic dependency with OutputWindowedValue).

I discussed this at length with @kennknowles and we decided to keep these setters and give up on moving this setup into the ProcessFn's constructor because there's no easy way to do this and because the set-up procedure of a ProcessFn is going to change in nearby PRs anyway (where I'll be adding more runner hooks). But the current PR, I think, is an acceptable intermediate state.